core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java

```diff
@@ -0,0 +1,191 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.page;
+
+import alluxio.client.file.cache.PageId;
+
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Specialized {@link PageId} when it's part of a block.
+ */
+public final class BlockPageId extends PageId {
+  // the name pattern of the page files of a block
+  // the block id and size are encoded as 16-byte hexadecimal
+  private static final Pattern FILE_ID_PATTERN =
+      Pattern.compile("paged_block_([0-9a-fA-F]{16})_size_([0-9a-fA-F]{16})");
+  private static final String FILE_ID_TEMPLATE = "paged_block_%016x_size_%016x";
+  // placeholder for temp blocks
+  private static final long INVALID_BLOCK_SIZE = -1;
+
+  /**
+   * this is constructed from {@link PageId#getFileId()} and cached to avoid parsing the string
+   * multiple times.
+   */
+  private final long mBlockId;
+
+  /**
+   * The block size. It is here because we need to put this information into the page store
+   * but it does not currently support page metadata.
+   * Todo(bowen): make this part of page metadata and remove it
+   */
+  private final long mBlockSize;
+
+  /**
+   * @param blockId the block ID
+   * @param pageIndex index of the page in the block
+   * @param blockSize block size
+   * @throws NumberFormatException when {@code blockId} cannot be parsed as a {@code long}
+   */
+  public BlockPageId(String blockId, long pageIndex, long blockSize) {
+    super(fileIdOf(Long.parseLong(blockId), blockSize), pageIndex);
+    mBlockId = Long.parseLong(blockId);
+    mBlockSize = blockSize;
+  }
+
+  /**
+   * Creates an instance with a block ID as a {@code long}.
+   * @param blockId the block ID
+   * @param pageIndex index of the page in the block
+   * @param blockSize block size
+   */
+  public BlockPageId(long blockId, long pageIndex, long blockSize) {
+    super(fileIdOf(blockId, blockSize), pageIndex);
+    mBlockId = blockId;
+    mBlockSize = blockSize;
+  }
+
+  /**
+   * Creates a new page of the temporary block.
+   * @param blockId
+   * @param pageIndex
+   * @return page ID
+   */
+  public static BlockPageId newTempPage(long blockId, long pageIndex) {
+    return new BlockPageId(blockId, pageIndex, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param blockId
+   * @param blockSize
+   * @return file ID
+   */
+  public static String fileIdOf(long blockId, long blockSize) {
+    return String.format(FILE_ID_TEMPLATE, blockId, blockSize).intern();
+  }
+
+  /**
+   * @param blockId
+   * @return file ID
+   */
+  public static String tempFileIdOf(long blockId) {
+    return fileIdOf(blockId, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param fileId
+   * @return block ID
+   */
+  public static Optional<Long> parseBlockId(String fileId) {
```

Comment:
in which situation we would expect empty? I feel like if there's no compatibility concern it should be runtime exception which indicate there must be something wrong if we can't parse the blockId